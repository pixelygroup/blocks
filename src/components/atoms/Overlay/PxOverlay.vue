<template>
  <transition name="fade">
    <div v-show="showOverlay" class="px-overlay" @click.stop="handleClick" />
  </transition>
</template>

<script>
export default {
  name: 'PxOverlay',

  props: {
    showOverlay: Boolean,
  },

  methods: {
    handleClick() {
      this.$emit('close')
    },
  },
}
</script>

<style lang="scss" scoped>
@import '../../../sass/framework/framework.scss';

.px-overlay {
  background-color: rgba(44, 62, 73, 0.55);
  bottom: 0;
  display: block;
  height: 100%;
  left: 0;
  position: fixed;
  right: 0;
  top: 0;
  width: 100%;
  z-index: $zindex-modal-overlay;
}

.fade-enter-active,
.fade-leave-active {
  transition: opacity 0.5s;
}

.fade-enter,
.fade-leave-to {
  opacity: 0;
}
</style>
