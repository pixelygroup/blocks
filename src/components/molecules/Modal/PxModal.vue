<template>
  <div>
    <px-overlay :show-overlay="show" @close="clickaway && closeModal()" />

    <transition
      name="fade"
      @before-enter="beforeEnterTransition"
      @after-leave="afterLeaveTransition"
    >
      <!--
        TODO: aria-labelledby
          This is tricky because it very much depends on what is passed into
          the slots by the parent. The ID use by the attribute must also be
          unique.
      -->
      <div
        v-if="show"
        class="px-modal"
        :class="[
          { 'px-modal--narrow': narrow, 'px-modal--wide': wide },
          `px-modal--${position}`,
        ]"
        :role="ariaRole"
        aria-modal="true"
        tabindex="-1"
        @click.self="clickaway && closeModal()"
      >
        <div class="px-modal__dialog" role="document" tabindex="0">
          <button
            v-if="showCloseButton"
            :aria-label="ariaLabel"
            data-qa-locator="modal-close-btn"
            type="button"
            class="px-modal__close"
            @click="closeModal"
          >
            <span class="sr-only">Close modal</span>
          </button>

          <!-- @slot Modal header -->
          <slot name="header"></slot>

          <!-- @slot Modal body -->
          <slot name="body"></slot>

          <!-- @slot Modal footer -->
          <slot name="footer"></slot>
        </div>
      </div>
    </transition>
  </div>
</template>

<script>
import PxOverlay from '../../atoms/Overlay/PxOverlay'

export default {
  name: 'PxModal',

  components: { PxOverlay },

  props: {
    /**
     * Whether to show the modal dialog
     */
    show: {
      type: Boolean,
      default: false,
    },
    /**
     * Whether to show the close button
     */
    showCloseButton: {
      type: Boolean,
      required: false,
      default: true,
    },
    /**
     * Whether the modal should be narrow (500px)
     */
    narrow: {
      type: Boolean,
      default: false,
    },
    /**
     * Whether the modal should be wide (700px)
     */
    wide: {
      type: Boolean,
      default: false,
    },
    /**
     * Position of the modal on the screen
     */
    position: {
      type: String,
      default: 'center',
      validator: value => ['top', 'bottom', 'center'].includes(value),
    },
    /**
     * Whether the modal can be dimissed by clicking on the overlay
     */
    clickaway: {
      type: Boolean,
      default: true,
    },

    /**
     * The modal ARIA role
     */
    ariaRole: {
      type: String,
      default: 'dialog',
      validator: value => ['dialog', 'alert', 'alertdialog'].includes(value),
    },

    /**
     * The modal button ARIA Label
     */
    ariaLabel: {
      type: String,
      default: 'Dismiss modal',
      required: false,
    },
  },

  data() {
    return {
      scrollbarWidth: 0,
      bodyElement: null,
      fixedElements: [],
      paddedElements: [],
    }
  },

  mounted() {
    this.bodyElement = document.body
    // TODO: find all fixed elements
    const siteHeader = document.querySelector('.m-site-header')

    siteHeader && this.fixedElements.push(siteHeader)
  },

  beforeDestroy() {
    // before the modal is unmounted, ensure that it tidies up after itself
    this.afterLeaveTransition()
  },

  methods: {
    /**
     * Sets up scrollbar and body before modal opens
     */
    beforeEnterTransition() {
      this.setScrollbarPadding()

      this.bodyElement.dataset.overflow = this.bodyElement.style.overflow
      this.bodyElement.classList.add('modal-open')
      this.bodyElement.style.overflow = 'hidden'
    },

    /**
     * Sets up scrollbar and body after modal closes
     */
    afterLeaveTransition() {
      const bodyOverflow = this.bodyElement.dataset.overflow

      if (typeof bodyOverflow === 'undefined' || bodyOverflow === '') {
        this.bodyElement.style.overflow = ''
      } else {
        this.bodyElement.style.bodyOverflow = bodyOverflow
        delete this.bodyElement.dataset.bodyOverflow
      }

      this.bodyElement.classList.remove('modal-open')
      this.resetScrollbarPadding()
    },

    /**
     * Emit the close modal event
     */
    closeModal() {
      /**
       * Close modal event
       *
       * @type {string}
       * @event 'close-modal'
       */
      this.$emit('close-modal')
    },

    /**
     * Gets the width of the browser's scrollbar
     *
     * @returns {number} the scrollbar width in pixels
     */
    getScrollbarWidth() {
      const scrollDiv = document.createElement('div')

      scrollDiv.style.cssText =
        'position: absolute; top: -9999px;width: 50px; height: 50px; overflow: scroll;'

      this.bodyElement.appendChild(scrollDiv)

      const scrollbarWidth =
        scrollDiv.getBoundingClientRect().width - scrollDiv.clientWidth

      this.bodyElement.removeChild(scrollDiv)

      return scrollbarWidth
    },

    /**
     * Check if the body is overflowing vertically
     *
     * @returns {boolean} whether the body is overflowing
     */
    isBodyOverflowing() {
      const rect = document.body.getBoundingClientRect()
      return rect.left + rect.right < window.innerWidth
    },

    /**
     * Sets the padding needed on elements if a scrollbar is present
     *
     * FIXME: if the modal isn't overflowing but the body is then the modal
     *  will appear too far to the right (by the scrollbar's width) due to
     *  there being no scrollbar on the overlay.
     * TODO: this currently doesn't consider position:sticky elements
     */
    setScrollbarPadding() {
      if (this.isBodyOverflowing()) {
        const scrollbarWidth = this.getScrollbarWidth()

        // Adjust the padding on any elements which are currently fixed position
        this.fixedElements.forEach(element => {
          if (window.getComputedStyle(element)['position'] === 'fixed') {
            this.setElementPadding(element, scrollbarWidth)
          }
        })

        // Adjust body padding
        this.setElementPadding(this.bodyElement, scrollbarWidth)
      }
    },

    /**
     * Get an element's padding
     *
     * @param {HTMLElement} element the element to get the padding from
     * @returns {object} the element's actual and calculated padding
     */
    getElementPadding(element) {
      const actualPadding = element.style.paddingRight
      const calculatedPadding = window.getComputedStyle(element)[
        'padding-right'
      ]

      return { actualPadding, calculatedPadding }
    },

    /**
     * Set an element's padding
     *
     * @param {HTMLElement} element the element on which to set the padding
     * @param {number} scrollbarWidth the scrollbar width
     */
    setElementPadding(element, scrollbarWidth) {
      const { actualPadding, calculatedPadding } = this.getElementPadding(
        element
      )

      element.dataset.paddingRight = actualPadding
      element.style.paddingRight = `${parseFloat(calculatedPadding) +
        scrollbarWidth}px`

      this.paddedElements.push(element)
    },

    /**
     * Reset the padding applied to element's for the scrollbar
     */
    resetScrollbarPadding() {
      this.paddedElements.forEach(element => this.resetElementPadding(element))
      this.paddedElements = []
    },

    /**
     * Reset an element's padding
     *
     * @param {HTMLElement} element the element on which to reset the padding
     */
    resetElementPadding(element) {
      const paddingRight = element.dataset.paddingRight

      if (typeof paddingRight === 'undefined') {
        element.style.paddingRight = ''
      } else {
        delete element.dataset.paddingRight
        element.style.paddingRight = paddingRight
      }
    },
  },
}
</script>

<style lang="scss" scoped>
@import '../../../sass/framework/framework.scss';

.fade-enter-active,
.fade-leave-active {
  transition: opacity 0.5s;
}

.fade-enter,
.fade-leave-to {
  opacity: 0;
}

.px-modal {
  bottom: 0;
  display: flex;
  flex-direction: column;
  left: 0;
  overflow-x: hidden;
  overflow-y: auto;
  position: fixed;
  right: 0;
  top: 0;
  z-index: $zindex-modal-overlay + 1;

  // @include breakpoint(xs-up) {
  //   padding: 30px 15px;
  // }

  // // This forces a 30px margin below the overflowed modal
  // @include breakpoint(xs-up) {
  //   &::after {
  //     content: '';
  //     display: block;
  //     flex-shrink: 0;
  //     height: 30px;
  //     width: 100%;
  //   }
  // }

  // &__dialog {
  //   background-color: #fff;
  //   border-radius: 0;
  //   flex: 1 0 auto;
  //   margin: auto;
  //   max-width: 500px;
  //   padding: 50px 40px;
  //   position: relative;

  //   @include breakpoint(xs-up) {
  //     border-radius: 10px;
  //     flex: 0 0 auto;
  //     width: 80vw;
  //   }

  //   @include breakpoint(sm-up) {
  //     padding: 50px 60px;
  //     width: 70vw;
  //   }

  //   @include breakpoint(md-up) {
  //     width: 60vw;
  //   }
  // }

  &--top {
    .modal__dialog {
      margin: 0 auto auto auto;
    }
  }

  &--bottom {
    .modal__dialog {
      margin: auto auto 0 auto;
    }
  }

  &--narrow {
    padding: 30px 15px;

    .modal__dialog {
      border-radius: 10px;
      flex: 0 0;
      margin-left: auto;
      margin-right: auto;
      max-width: 350px;
    }
  }

  &--wide {
    .px-modal__dialog {
      max-width: $container-xs;
    }
  }

  &__close {
    background-color: #fff;
    border: none;
    color: $color-primary;
    cursor: pointer;
    font-size: 1.5em;
    height: 30px;
    padding: 0;
    position: absolute;
    right: 15px;
    top: 15px;
    transition: scale 0.5s;
    width: 30px;

    &::after {
      content: '\2715';
      display: block;
    }

    &:hover {
      scale: 1.2;
    }
  }
}
</style>
