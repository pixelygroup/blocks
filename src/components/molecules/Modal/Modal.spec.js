import { mount, shallowMount } from '@vue/test-utils'
import PxModal from './PxModal'
import PxOverlay from '../../atoms/Overlay/PxOverlay'

const modalSelector = '.px-modal'
const modalCloseButtonSelector = '.px-modal__close'
const modalDialogSelector = '.px-modal__dialog'

describe('Modal', () => {
  test('renders correctly', () => {
    const wrapper = shallowMount(PxModal, {
      propsData: {
        show: true,
      },
      slots: {
        header: '<p>This is the header</p>',
        body: '<p>This is the body</p>',
        footer: '<p>This is the body text</p>',
      },
    })
    expect(wrapper).toMatchSnapshot()
  })

  test('the modal is initially hidden', () => {
    const wrapper = mount(PxModal)
    expect(wrapper.props().show).toBe(false)

    expect(wrapper.find(modalSelector).exists()).toBeFalsy()
    expect(wrapper.find(PxOverlay).isVisible()).toBeFalsy()
  })

  test('should display the modal on show=true', () => {
    const wrapper = mount(PxModal)

    wrapper.setProps({ show: true })

    expect(wrapper.find(modalSelector).exists()).toBeTruthy()
    expect(wrapper.find(PxOverlay).isVisible()).toBeTruthy()
  })

  test('should hide the modal on show=false', () => {
    const wrapper = mount(PxModal, {
      propsData: {
        show: true,
      },
    })

    expect(wrapper.find(modalSelector).exists()).toBeTruthy()
    expect(wrapper.find(PxOverlay).isVisible()).toBeTruthy()

    wrapper.setProps({ show: false })

    expect(wrapper.find(modalSelector).exists()).toBeFalsy()
    expect(wrapper.find(PxOverlay).isVisible()).toBeFalsy()
  })

  test('the close button can be hidden', () => {
    const wrapper = shallowMount(PxModal, {
      propsData: {
        show: true,
        showCloseButton: false,
      },
    })

    expect(wrapper.find(modalCloseButtonSelector).exists()).toBeFalsy()
  })

  test('the default position is center', () => {
    const wrapper = shallowMount(PxModal, {
      propsData: {
        show: true,
      },
    })

    expect(wrapper.find(modalSelector).classes()).toContain('px-modal--center')
  })

  test('the position can be changed', () => {
    const wrapper = shallowMount(PxModal, {
      propsData: {
        show: true,
        position: 'top',
      },
    })

    expect(wrapper.find(modalSelector).classes()).toContain('px-modal--top')
    expect(wrapper.find(modalSelector).classes()).not.toContain(
      'modal--center modal--bottom'
    )

    wrapper.setProps({ position: 'bottom' })

    expect(wrapper.find(modalSelector).classes()).toContain('px-modal--bottom')
    expect(wrapper.find(modalSelector).classes()).not.toContain(
      'modal--center modal--top'
    )
  })

  test('the modal width can be restricted', () => {
    const wrapper = shallowMount(PxModal, {
      propsData: {
        show: true,
        narrow: true,
      },
    })

    expect(wrapper.find(modalSelector).classes()).toContain('px-modal--narrow')
  })

  test('the aria role can be changed', () => {
    const wrapper = shallowMount(PxModal, {
      propsData: {
        show: true,
      },
    })

    // Default value
    expect(wrapper.find(modalSelector).attributes().role).toEqual('dialog')

    wrapper.setProps({ ariaRole: 'alert' })
    expect(wrapper.find(modalSelector).attributes().role).toEqual('alert')

    wrapper.setProps({ ariaRole: 'alertdialog' })
    expect(wrapper.find(modalSelector).attributes().role).toEqual('alertdialog')

    wrapper.setProps({ ariaRole: 'dialog' })
    expect(wrapper.find(modalSelector).attributes().role).toEqual('dialog')
  })

  describe('modal close event', () => {
    test('clicking on the modal backdrop emit the close event', () => {
      const wrapper = shallowMount(PxModal, {
        propsData: {
          show: true,
        },
      })

      wrapper.find(modalSelector).trigger('click')
      expect(wrapper.emitted('close-modal')).toBeTruthy()
    })

    test('clicking on the modal dialog does not emit the close event', () => {
      const wrapper = shallowMount(PxModal, {
        propsData: {
          show: true,
        },
      })

      wrapper.find(modalDialogSelector).trigger('click')
      expect(wrapper.emitted('close-modal')).toBeFalsy()
    })

    test('clicking on the modal close button emits the close event', () => {
      const wrapper = shallowMount(PxModal, {
        propsData: {
          show: true,
        },
      })

      wrapper.find(modalCloseButtonSelector).trigger('click')
      expect(wrapper.emitted('close-modal')).toBeTruthy()
    })

    test('disabling clickaway prevents emitting the close event on backdrop click', () => {
      const wrapper = shallowMount(PxModal, {
        propsData: {
          show: true,
          clickaway: false,
        },
      })

      wrapper.find(modalSelector).trigger('click')
      expect(wrapper.emitted('close-modal')).toBeFalsy()
    })
  })

  /**FIXME: JSDOM currently stubs some layout apis to return dummy values which,
  /*  without defining all the various missing properties on element, makes this
  /*  test unable to be completed.
  /* https://github.com/jsdom/jsdom#unimplemented-parts-of-the-web-platform
  **/
  /* eslint-disable jest/no-disabled-tests */
  describe.skip('scrollbar behaviour', () => {
    test.skip('when open, if the body has overflow, the body overflow is hidden and padding which equals the width of the scrollbar added to the body or any fixed/sticky element', () => {})

    test.skip('when closed, the body overflow/padding and any fixed/sticky elements padding are restored to what they were before the modal opened', () => {})
  })
  /* eslint-enable jest/no-disabled-tests */
})
