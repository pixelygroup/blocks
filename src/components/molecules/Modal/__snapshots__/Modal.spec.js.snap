// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders correctly 1`] = `
<div>
  <px-overlay-stub showoverlay="true"></px-overlay-stub>
  <div role="dialog" aria-modal="true" tabindex="-1" class="px-modal px-modal--center" name="fade">
    <div role="document" tabindex="0" class="px-modal__dialog"><button aria-label="Dismiss modal" data-qa-locator="modal-close-btn" type="button" class="px-modal__close"><span class="sr-only">Close modal</span></button>
      <p>This is the header</p>
      <p>This is the body</p>
      <p>This is the body text</p>
    </div>
  </div>
</div>
`;
